src/psbt/mod.rs

```diff
@@ -363,35 +383,586 @@ fn sanity_check(psbt: &Psbt) -> Result<(), Error> {
+                .ecdsa_hash_ty()
+                .map_err(|e| Error::InputError(InputError::NonStandardSigHashType(e), index))?,
+            None => EcdsaSigHashType::All,
+        };
+        for (key, ecdsa_sig) in &input.partial_sigs {
+            let flag = bitcoin::EcdsaSigHashType::from_u32_standard(ecdsa_sig.hash_ty as u32)
+                .map_err(|_| {
+                    Error::InputError(
+                        InputError::Interpreter(interpreter::Error::NonStandardSigHash(
+                            ecdsa_sig.to_vec(),
+                        )),
+                        index,
+                    )
+                })?;
+            if target_ecdsa_sighash_ty != flag {
+                return Err(Error::InputError(
+                    InputError::WrongSigHashFlag {
+                        required: target_ecdsa_sighash_ty,
+                        got: flag,
+                        pubkey: bitcoin::PublicKey::new(*key),
+                    },
+                    index,
+                ));
+            }
+            // Signatures are well-formed in psbt partial sigs
+        }
+    }
+
     Ok(())
 }
 
-/// Psbt extractor as defined in BIP174 that takes in a psbt reference
-/// and outputs a extracted bitcoin::Transaction
-/// Also does the interpreter sanity check
-/// Will error if the final ScriptSig or final Witness are missing
-/// or the interpreter check fails.
-pub fn extract<C: secp256k1::Verification>(
-    psbt: &Psbt,
-    secp: &Secp256k1<C>,
-) -> Result<bitcoin::Transaction, Error> {
-    sanity_check(psbt)?;
-
-    let mut ret = psbt.unsigned_tx.clone();
-    for (n, input) in psbt.inputs.iter().enumerate() {
-        if input.final_script_sig.is_none() && input.final_script_witness.is_none() {
-            return Err(Error::InputError(InputError::MissingWitness, n));
+/// Additional operations for miniscript descriptors for various psbt roles.
+/// Note that these APIs would generally error when used on scripts that are not
+/// miniscripts.
+pub trait PsbtExt {
+    /// Finalize the psbt. This function takes in a mutable reference to psbt
+    /// and populates the final_witness and final_scriptsig
+    /// for all miniscript inputs.
+    ///
+    /// Finalizes all inputs that it can finalize, and returns an error for each input
+    /// that it cannot finalize. Also performs a sanity interpreter check on the
+    /// finalized psbt which involves checking the signatures/ preimages/timelocks.
+    ///
+    /// Input finalization also fails if it is not possible to satisfy any of the inputs non-malleably
+    /// See [finalizer::finalize_mall] if you want to allow malleable satisfactions
+    ///
+    /// For finalizing individual inputs, see also [`PsbtExt::finalize_inp`]
+    fn finalize_mut<C: secp256k1::Verification>(
+        &mut self,
+        secp: &secp256k1::Secp256k1<C>,
+    ) -> Result<(), Vec<Error>>;
+
+    /// Same as [`PsbtExt::finalize_mut`], but does not mutate the input psbt and
```

Comment:
doc issue: it does mutate the input, it owns it?

what's different here is that it's clear that unless you're the owner of the psbt, you don't clobber it and return the psbt either in the Ok or Err.